src/module.c

```diff
@@ -5271,10 +5271,16 @@ int RM_HashSet(RedisModuleKey *key, int flags, ...) {
 
         /* Handle XX and NX */
         if (flags & (REDISMODULE_HASH_XX|REDISMODULE_HASH_NX)) {
-            int isHashDeleted;
-            int exists = hashTypeExists(key->db, key->value, field->ptr, &isHashDeleted);
-            /* hash-field-expiration is not exposed to modules */
-            serverAssert(isHashDeleted == 0);
+            int hfeFlags = HFE_NO_HASH_EXPIRE; /* Avoid invalidate the key */
+
+            /* This is a bit tricky: hash might hold expired field. If we lazy
+             * delete it, and command was sent with XX, then operation
+             * will fail and hash will be left wrongly empty. For this case, let's
+             * avoid lazy delete (and let this operation to fail) */
```

Comment:
Not sure if this comment is accurate. If key becomes empty, it will be deleted at the end of this while loop, right? (with moduleDelKeyIfEmpty() call). So, it is not possible to leave it as empty.

Reply:
I am not that experienced with modules but the command `RM_HashSet` is a bundle of few flavours of hash modification. At some flavours the caller expects that the hash might be deleted underneath, whereas other operations, the caller will less expect it to happen. 

Most of the users will write module code that try to set a field without expecting to a side effect of deleted hash in their context and in turn to invaildate their key handle. To be on the safe side, I think we should just leave an expired field. WDYT?

Reply:
I think the code is fine here. 

I just wanted to point that comment was slightly inaccurate as leaving the key empty is not possible. So, you may say, key would be deleted instead of saying it would be left empty.

Reply:
Note that moduleDelKeyIfEmpty() won't delete the hash if it left with single expired key because hash counts blindly expired fields as well. I modified the comment. please see if it makes sense to you. 
Thanks.